gpbft/gpbft.go

```diff
@@ -294,90 +299,90 @@ func (i *instance) Describe() string {
 			ErrValidationWrongSupplement, msg.Vote.SupplementalData, i.supplementalData)
 	}
 	// Check proposal has the expected base chain.
 	if !(msg.Vote.Value.IsZero() || msg.Vote.Value.HasBase(i.input.Base())) {
-		return fmt.Errorf("%w: message base %s, expected %s",
+		return false, fmt.Errorf("%w: message base %s, expected %s",
 			ErrValidationWrongBase, &msg.Vote.Value, i.input.Base())
 	}
 
 	if i.phase == TERMINATED_PHASE {
-		return nil // No-op
+		return false, nil // No-op
 	}
 	// Ignore QUALITY messages after exiting the QUALITY phase.
 	// Ignore CONVERGE and PREPARE messages for prior rounds.
 	forPriorRound := msg.Vote.Round < i.round
 	if (msg.Vote.Step == QUALITY_PHASE && i.phase != QUALITY_PHASE) ||
 		(forPriorRound && msg.Vote.Step == CONVERGE_PHASE) ||
 		(forPriorRound && msg.Vote.Step == PREPARE_PHASE) {
-		return nil
+		return false, nil
 	}
 
 	// Drop message that:
 	//  * belong to future rounds, beyond the configured max lookahead threshold, and
 	//  * carry no justification, i.e. are spammable.
 	beyondMaxLookaheadRounds := msg.Vote.Round > i.round+i.participant.maxLookaheadRounds
 	if beyondMaxLookaheadRounds && isSpammable(msg) {
-		return nil
+		return false, nil
 	}
 
-	round := i.roundState(msg.Vote.Round)
+	// Load the round state and process further only valid, non-spammable messages.
+	// Equivocations are handled by the quorum state.
+	msgRound := i.getRound(msg.Vote.Round)
 	switch msg.Vote.Step {
 	case QUALITY_PHASE:
 		// Receive each prefix of the proposal independently.
 		i.quality.ReceiveEachPrefix(msg.Sender, msg.Vote.Value)
 	case CONVERGE_PHASE:
-		if err := round.converged.Receive(msg.Sender, msg.Vote.Value, msg.Ticket, msg.Justification); err != nil {
-			return fmt.Errorf("failed processing CONVERGE message: %w", err)
+		if err := msgRound.converged.Receive(msg.Sender, msg.Vote.Value, msg.Ticket, msg.Justification); err != nil {
+			return false, fmt.Errorf("failed processing CONVERGE message: %w", err)
 		}
 	case PREPARE_PHASE:
-		round.prepared.Receive(msg.Sender, msg.Vote.Value, msg.Signature)
+		msgRound.prepared.Receive(msg.Sender, msg.Vote.Value, msg.Signature)
 	case COMMIT_PHASE:
-		round.committed.Receive(msg.Sender, msg.Vote.Value, msg.Signature)
+		msgRound.committed.Receive(msg.Sender, msg.Vote.Value, msg.Signature)
 		// The only justifications that need to be stored for future propagation are for COMMITs
 		// to non-bottom values.
 		// This evidence can be brought forward to justify a CONVERGE message in the next round.
 		if !msg.Vote.Value.IsZero() {
-			round.committed.ReceiveJustification(msg.Vote.Value, msg.Justification)
+			msgRound.committed.ReceiveJustification(msg.Vote.Value, msg.Justification)
 		}
 	case DECIDE_PHASE:
 		i.decision.Receive(msg.Sender, msg.Vote.Value, msg.Signature)
 		if i.phase != DECIDE_PHASE {
 			i.skipToDecide(msg.Vote.Value, msg.Justification)
 		}
-		if err := i.tryDecide(); err != nil {
-			return fmt.Errorf("failed to decide: %w", err)
-		}
```

Comment:
This is redundant after implementation of skipToDecide, immediately above. `tryDecide()` will be invoked by `tryCurrentPhase()`  at the end.